spec/models/users_role_spec.rb

```diff
@@ -8,31 +8,81 @@
 #
 RSpec.describe UsersRole, type: :model do
   describe "#current_role_for" do
-    context "for org user" do
-      it "should return org user" do
-        user = FactoryBot.create(:user)
-        expect(described_class.current_role_for(user).name).to eq("org_user")
+    context "when last_role is nil" do
+      context "for org user" do
+        it "should return org user" do
+          user = create(:user)
+          expect(UsersRole.current_role_for(user).name).to eq("org_user")
+        end
+      end
+
+      context "for org admin" do
+        it "should return org admin" do
+          user = create(:organization_admin)
+          expect(UsersRole.current_role_for(user).name).to eq("org_admin")
+        end
+      end
+
+      context "for super admin" do
+        it "should return super admin" do
+          user = create(:super_admin)
+          expect(UsersRole.current_role_for(user).name).to eq("super_admin")
+        end
+      end
+
+      context "for partner user" do
+        it "should return partner user" do
+          user = create(:partner_user)
+          expect(UsersRole.current_role_for(user).name).to eq("partner")
+        end
       end
     end
+    context "when last_role is not nil" do
+      it "should return last role" do
+        user = create(:partner_user)
 
-    context "for org admin" do
-      it "should return org admin" do
-        user = FactoryBot.create(:organization_admin)
-        expect(described_class.current_role_for(user).name).to eq("org_admin")
+        UsersRole.set_last_role_for(user, user.roles.last)
+
+        user.add_role(Role::ORG_ADMIN, create(:organization))
+        expect(UsersRole.current_role_for(user).name).to eq("partner")
       end
     end
+  end
 
-    context "for super admin" do
-      it "should return super admin" do
-        user = FactoryBot.create(:super_admin)
-        expect(described_class.current_role_for(user).name).to eq("super_admin")
+  describe "#set_last_role_for" do
+    context "when user has the role" do
+      it "should set last role" do
+        user = create(:partner_user)
+        role = user.roles.first
+
+        got = UsersRole.set_last_role_for(user, role)
+
+        expect(user.last_role).to eq(role)
+        expect(got).to be true
+      end
+
+      # TODO: I don't know the best place for this test
+      it "should remove last role if users role is removed" do
+        user = create(:partner_user)
+        role = user.roles.first
+
+        UsersRole.set_last_role_for(user, role)
+        expect(user.last_role).to eq(role)
+
+        RemoveRoleService.call(user_id: user.id, role_id: role.id)
+
+        expect(user.reload.last_role).to eq(nil)
```

Comment:
Not sure if the is the best place for this test. Its basically testing that the association is set up correctly but it is critical to check that last role isn't storing a revoked role.

Reply:
Yeah this should be part of the `RemoveRoleService` test.